modules/standard/IO.chpl

```diff
@@ -3838,6 +4098,48 @@ proc _channel._constructIoErrorMsg(param kind: iokind, const x:?t): string {
   return result;
 }
 
+proc _channel._decodeOne(type readType, loc:locale) throws {
+  var reader = new fileReader(iokind.dynamic, locking=false,
+                              formatter=_fmt,
+                              home=here,
+                              _channel_internal=_channel_internal,
+                              _readWriteThisFromLocale=loc);
+  defer { reader._channel_internal = QIO_CHANNEL_PTR_NULL; }
+
+  if isClassType(readType) {
+    // Save formatter authors from having to reason about 'owned' and
+    // 'shared' by converting the type to unmanaged.
+    var tmp = reader.formatter.decode(reader, _to_unmanaged(readType));
+
+    if isOwnedClassType(readType) {
+      return new _owned(tmp);
+    } else if isSharedClassType(readType) {
+      return new _shared(tmp);
```

Comment:
We want the above code to work also for user-defined management types. I recommend adding a TODO comment for now. Eventually, I think it will be `if !isUnmanagedClassType(readType) && !isBorrowedClassType(readType)` and then `return readType.adopt(tmp);`. BTW, can you decode a `borrowed` class instance? It looks like this code would not handle it, so that's probably also worth a TODO comment.